instrumentation/kafka/kafka-clients/kafka-clients-common/library/src/main/java/io/opentelemetry/instrumentation/kafkaclients/OpenTelemetryKafkaMetrics.java

```diff
@@ -120,86 +70,58 @@ public void init(List<KafkaMetric> metrics) {
       logger.log(Level.FINEST, "Metric changed but meter not set: {0}", kafkaMetricId);
       return;
     }
 
-    InstrumentDescriptor instrumentDescriptor =
-        KafkaMetricRegistry.getInstrumentDescriptor(kafkaMetricId);
-    if (instrumentDescriptor == null) {
+    RegisteredObservable registeredObservable =
+        KafkaMetricRegistry.getRegisteredObservable(currentMeter, kafkaMetricId, metric);
+    if (registeredObservable == null) {
       logger.log(Level.FINEST, "Metric changed but cannot map to instrument: {0}", kafkaMetricId);
       return;
     }
 
-    RegisteredInstrument registeredInstrument =
-        RegisteredInstrument.create(kafkaMetricId, toAttributes(metric));
-
-    instrumentMap.compute(
-        registeredInstrument,
-        (registeredInstrument1, autoCloseable) -> {
-          if (autoCloseable != null) {
-            logger.log(Level.FINEST, "Replacing instrument {0}", registeredInstrument1);
-            try {
-              autoCloseable.close();
-            } catch (Exception e) {
-              logger.log(Level.WARNING, "An error occurred closing instrument", e);
-            }
-          } else {
-            logger.log(Level.FINEST, "Adding instrument {0}", registeredInstrument1);
-          }
-          return createObservable(
-              currentMeter, registeredInstrument1, instrumentDescriptor, metric);
-        });
-  }
-
-  private static AutoCloseable createObservable(
-      Meter meter,
-      RegisteredInstrument registeredInstrument,
-      InstrumentDescriptor instrumentDescriptor,
-      KafkaMetric kafkaMetric) {
-    Consumer<ObservableDoubleMeasurement> callback =
-        observableMeasurement ->
-            observableMeasurement.record(kafkaMetric.value(), registeredInstrument.getAttributes());
-    switch (instrumentDescriptor.getInstrumentType()) {
-      case INSTRUMENT_TYPE_DOUBLE_OBSERVABLE_GAUGE:
-        return meter
-            .gaugeBuilder(instrumentDescriptor.getName())
-            .setDescription(instrumentDescriptor.getDescription())
-            .buildWithCallback(callback);
-      case INSTRUMENT_TYPE_DOUBLE_OBSERVABLE_COUNTER:
-        return meter
-            .counterBuilder(instrumentDescriptor.getName())
-            .setDescription(instrumentDescriptor.getDescription())
-            .ofDoubles()
-            .buildWithCallback(callback);
-      default: // Continue below to throw
+    InstrumentDescriptor instrumentDescriptor = registeredObservable.getInstrumentDescriptor();
+    Attributes attributes = registeredObservable.getAttributes();
+    Set<AttributeKey<?>> attributeKeys = attributes.asMap().keySet();
+
+    for (Iterator<RegisteredObservable> it = registeredObservables.iterator(); it.hasNext(); ) {
+      RegisteredObservable curRegisteredObservable = it.next();
+      Set<AttributeKey<?>> curAttributeKeys =
+          curRegisteredObservable.getAttributes().asMap().keySet();
+      if (curRegisteredObservable.equals(registeredObservable)) {
```

Comment:
Should it be `equals()` here? The `AutoClosable`/`SdkObservableInstrument` does not implement `equals()`

Reply:
Good catch! I made `RegisteredObservable#matches(KafkaMetric)` for this purpose but forgot to use it here. But on closer inspection, even that isn't quite as good as it could be. I actually just want to look for the `RegisteredObserver` associated with the same `MetricName` of the `KafkaMetric`. `MetricName` defines the attributes of distinct metric series, consisting of name, group, description, and tags. It has a proper `equals` implementation too so is exactly what is needed here.